Automated cherry pick of #38818

gnufied opened the pull request:
Cherry pick of #38818 on release-1.4.

jsafrane committed:
AWS: Add sequential allocator for device names.

On AWS, we should not reuse device names as long as possible, see
https://aws.amazon.com/premiumsupport/knowledge-center/ebs-stuck-attaching/

"If you specify a device name that is not in use by EC2, but is being used by
the block device driver within the EC2 instance, the attachment of the EBS
volume does not succeed and the EBS volume is stuck in the attaching state."

This patch adds a device name allocator that tries to find a name that's next
to the last used device name instead of using the first available one.
This way we will loop through all device names ("xvdba" .. "xvdzz") before
a device name is reused.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40399)
<!-- Reviewable:end -->

gnufied committed:
Change type of deviceAllocator map to string

types.NodeName doesn't exist in 1.4 branch. AWS resources
were strongly typed as part of
https://github.com/kubernetes/kubernetes/pull/35883
but merging that PR is even trickier.
